sample.go

```diff
@@ -0,0 +1,215 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package environschema
+
+import (
+	"bytes"
+	"fmt"
+	"go/doc"
+	"io"
+	"reflect"
+	"sort"
+	"strings"
+	"unicode"
+
+	"gopkg.in/yaml.v2"
+)
+
+// SampleYAML writes YAML output to w, indented by indent spaces
+// that holds the attributes in attrs with descriptions found
+// in the given fields. An entry for any attribute in fields not
+// in attrs will be generated but commented out.
+func SampleYAML(w io.Writer, indent int, attrs map[string]interface{}, fields Fields) error {
+	indentStr := strings.Repeat(" ", indent)
+	orderedFields := make(fieldsByGroup, 0, len(fields))
+	for name, f := range fields {
+		orderedFields = append(orderedFields, attrWithName{
+			name: name,
+			Attr: f,
+		})
+	}
+	sort.Sort(orderedFields)
+	for i, f := range orderedFields {
+		if i > 0 {
+			fmt.Fprintf(w, "\n")
+		}
+		writeSampleDescription(w, f.Attr, indentStr+"# ")
+		val, ok := attrs[f.name]
+		if ok {
+			fmt.Fprintf(w, "%s:", f.name)
+			indentVal(w, val, indentStr)
+		} else {
+			if f.Example != nil {
+				val = f.Example
+			} else {
+				val = sampleValue(f.Type)
+			}
+			fmt.Fprintf(w, "# %s:", f.name)
+			indentVal(w, val, indentStr+"# ")
+		}
+	}
+	return nil
+}
+
+const textWidth = 80
+
+func writeSampleDescription(w io.Writer, f Attr, indent string) {
+	previousText := false
+
+	// section marks the start of a new section of the comment;
+	// sections are separated with empty lines.
+	section := func() {
+		if previousText {
+			fmt.Fprintf(w, "%s\n", strings.TrimRightFunc(indent, unicode.IsSpace))
+		}
+		previousText = true
+	}
+
+	descr := strings.TrimSpace(f.Description)
+	if descr != "" {
+		section()
+		if !strings.HasSuffix(descr, ".") {
```

Comment:
Are you sure you want to do this? Sentences might end with a ! or a ? (although that might be unusual for a description)